# Natural Language Processing Project

Bagging Model for Product Title Quality with Noise

CIKM AnalytiCup 2017

Aarib Ahmed Vahidy

Partham Kumar Chawla

Arham Hussain Khan

### Loading Libraries

In [1]:
pip install pandas numpy scikit-learn xgboost lightgbm nltk gensim beautifulsoup4

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
import numpy as np
import re
import string
import nltk
from sklearn.model_selection import KFold
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import Ridge, LogisticRegression
from sklearn.ensemble import RandomForestRegressor, ExtraTreesRegressor
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import OneHotEncoder
import lightgbm as lgb
import xgboost as xgb
import warnings
from bs4 import BeautifulSoup
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import mean_squared_error
from tqdm import tqdm

warnings.filterwarnings('ignore')

### Data Loading and Label Integration

Training and validation datasets are loaded from CSV files.

Corresponding labels (clarity and conciseness) are loaded from separate .labels and .predict files.

These labels are assigned meaningful column names.

The label files are then concatenated with their respective main datasets.

Column names of both train and validation DataFrames are standardized for consistency and easier downstream processing.

### Loading training data

In [3]:
#Loading the main data
data_train = pd.read_csv(r'C:\Users\aarib\7thSemester\NLP\NLPProject\CIKMAnalytiCup2017_Lazada\training\data_train.csv')

#Loading labels
clarity_labels = pd.read_csv(r'C:\Users\aarib\7thSemester\NLP\NLPProject\CIKMAnalytiCup2017_Lazada\training\clarity_train.labels', header=None)
conciseness_labels = pd.read_csv(r'C:\Users\aarib\7thSemester\NLP\NLPProject\CIKMAnalytiCup2017_Lazada\training\conciseness_train.labels', header=None)

#Assigning meaningful column names
clarity_labels.columns = ['clarity']
conciseness_labels.columns = ['conciseness']

#Combining everything into one DataFrame
train = pd.concat([data_train, clarity_labels, conciseness_labels], axis=1)

print("Shape of training data:", train.shape)
train.head()

Shape of training data: (36283, 11)


,my,AD674FAASTLXANMY,Adana Gallery Suri Square Hijab – Light Pink,Fashion,Women,Muslim Wear,<ul><li>Material : Non sheer shimmer chiffon</li><li>Sizes : 52 x 52 inches OR 56 x 56 inches</li><li>Cut with curved ends</li></ul>,49.0,local,clarity,conciseness
0,my,AE068HBAA3RPRDANMY,Cuba Heartbreaker Eau De Parfum Spray 100ml/3.3oz,Health & Beauty,Bath & Body,Hand & Foot Care,Formulated with oil-free hydrating botanicals/...,128.00,international,1,1
1,my,AN680ELAA9VN57ANMY,Andoer 150cm Cellphone Smartphone Mini Dual-He...,"TV, Audio / Video, Gaming & Wearables",Audio,Live Sound & Stage,<ul> <li>150cm mini microphone compatible for ...,25.07,international,1,1
2,my,AN957HBAAAHDF4ANMY,ANMYNA Complaint Silky Set 柔顺洗发配套 (Shampoo 520...,Health & Beauty,Hair Care,Shampoos & Conditioners,<ul> <li>ANMYNA Complaint Silky Set (Shampoo 5...,118.00,local,1,0
3,my,AR511HBAXNWAANMY,Argital Argiltubo Green Clay For Face and Body...,Health & Beauty,Men's Care,Body and Skin Care,<ul> <li>100% Authentic</li> <li>Rrefresh and ...,114.80,international,1,1
4,my,AS575ELCMZ4WANMY,Asus TP300LJ-DW004H Transformer Book Flip 4GB ...,Computers & Laptops,Laptops,Traditional Laptops,"<div class=""prod_content""> <div class=""prod_de...",2599.00,local,1,1


### Adding column names to training data

In [4]:
train.columns = [
    'country', 'product_id', 'title', 
    'category_lvl_1', 'category_lvl_2', 'category_lvl_3',
    'description', 'price', 'delivery_type', 
    'clarity', 'conciseness'
]

In [5]:
train.head()

,country,product_id,title,category_lvl_1,category_lvl_2,category_lvl_3,description,price,delivery_type,clarity,conciseness
0,my,AE068HBAA3RPRDANMY,Cuba Heartbreaker Eau De Parfum Spray 100ml/3.3oz,Health & Beauty,Bath & Body,Hand & Foot Care,Formulated with oil-free hydrating botanicals/...,128.00,international,1,1
1,my,AN680ELAA9VN57ANMY,Andoer 150cm Cellphone Smartphone Mini Dual-He...,"TV, Audio / Video, Gaming & Wearables",Audio,Live Sound & Stage,<ul> <li>150cm mini microphone compatible for ...,25.07,international,1,1
2,my,AN957HBAAAHDF4ANMY,ANMYNA Complaint Silky Set 柔顺洗发配套 (Shampoo 520...,Health & Beauty,Hair Care,Shampoos & Conditioners,<ul> <li>ANMYNA Complaint Silky Set (Shampoo 5...,118.00,local,1,0
3,my,AR511HBAXNWAANMY,Argital Argiltubo Green Clay For Face and Body...,Health & Beauty,Men's Care,Body and Skin Care,<ul> <li>100% Authentic</li> <li>Rrefresh and ...,114.80,international,1,1
4,my,AS575ELCMZ4WANMY,Asus TP300LJ-DW004H Transformer Book Flip 4GB ...,Computers & Laptops,Laptops,Traditional Laptops,"<div class=""prod_content""> <div class=""prod_de...",2599.00,local,1,1


### Loading validation data

In [6]:
#Loading the main data
data_val = pd.read_csv(r'C:\Users\aarib\7thSemester\NLP\NLPProject\CIKMAnalytiCup2017_Lazada\validation\data_valid.csv')

#Loading labels
clarity_labels = pd.read_csv(r'C:\Users\aarib\7thSemester\NLP\NLPProject\CIKMAnalytiCup2017_Lazada\validation\clarity_valid.predict', header=None)
conciseness_labels = pd.read_csv(r'C:\Users\aarib\7thSemester\NLP\NLPProject\CIKMAnalytiCup2017_Lazada\validation\conciseness_valid.predict', header=None)

#Assigning meaningful column names
clarity_labels.columns = ['clarity']
conciseness_labels.columns = ['conciseness']

#Combining everything into one DataFrame
validation = pd.concat([data_val, clarity_labels, conciseness_labels], axis=1)

print("Shape of validation data:", validation.shape)
validation.head()

Shape of validation data: (11838, 11)


,my,AP564ELASSTWANMY,Apple MacBook Pro MGXC2ZP/A 16GB i7 15.4-inch Retina Display Laptop,Computers & Laptops,Laptops,Macbooks,OS X Lion<br> Intel Core i7<br> 15-inch Retina Display<br> 16GB RAM / 512GB Flash<br> GeForce GT 750M<br> WiFi + BT4 + RJ45<br>,12550.0,local,clarity,conciseness
0,my,BR924HBAA5B3TLANMY,BRAND'S® American Ginseng Triple Pack (3x 6's)...,Health & Beauty,Food Supplements,Well Being,<ul> <li>Traditionally used to calm the mind a...,105.0,local,0.34139,0.98747
1,my,CA673ELAA5UG3XANMY,Canon EOS M10 Mirrorless Digital Camera 18MP w...,Cameras,Mirrorless,NaN,<div> <ul> <li>18.0MP APS-C CMOS Sensor</li> <...,1588.0,local,0.22758,0.29770
2,my,DE759ELAA7QM1XANMY,"Dell LED Monitor 23"" (E2316H)",Computers & Laptops,Computer Accessories,Monitors,"<div class=""prod_content""> <div class=""prod_de...",565.0,local,0.36162,0.59796
3,my,ES802OTAABHAY8ANMY,Esprit Tallac Brave Nubuck Sand ES107601001 Be...,Watches Sunglasses Jewellery,Watches,Men,<ul> <li>stainless steel case</li> <li>mineral...,279.0,local,0.83418,0.56371
4,my,HP961ELAABF7N7ANMY,"(Refurbished) HP Compaq 3330 Pro MT + 19"" LCD",Computers & Laptops,Desktops Computers,All-purpose,"<ul> <li>Model : HP Compaq 3330 Pro MT + 19"" L...",1259.0,local,0.12018,0.50220


### Adding column names to validation data

In [7]:
validation.columns = [
    'country', 'product_id', 'title', 
    'category_lvl_1', 'category_lvl_2', 'category_lvl_3',
    'description', 'price', 'delivery_type', 
    'clarity', 'conciseness'
]

In [8]:
validation.head()

,country,product_id,title,category_lvl_1,category_lvl_2,category_lvl_3,description,price,delivery_type,clarity,conciseness
0,my,BR924HBAA5B3TLANMY,BRAND'S® American Ginseng Triple Pack (3x 6's)...,Health & Beauty,Food Supplements,Well Being,<ul> <li>Traditionally used to calm the mind a...,105.0,local,0.34139,0.98747
1,my,CA673ELAA5UG3XANMY,Canon EOS M10 Mirrorless Digital Camera 18MP w...,Cameras,Mirrorless,NaN,<div> <ul> <li>18.0MP APS-C CMOS Sensor</li> <...,1588.0,local,0.22758,0.29770
2,my,DE759ELAA7QM1XANMY,"Dell LED Monitor 23"" (E2316H)",Computers & Laptops,Computer Accessories,Monitors,"<div class=""prod_content""> <div class=""prod_de...",565.0,local,0.36162,0.59796
3,my,ES802OTAABHAY8ANMY,Esprit Tallac Brave Nubuck Sand ES107601001 Be...,Watches Sunglasses Jewellery,Watches,Men,<ul> <li>stainless steel case</li> <li>mineral...,279.0,local,0.83418,0.56371
4,my,HP961ELAABF7N7ANMY,"(Refurbished) HP Compaq 3330 Pro MT + 19"" LCD",Computers & Laptops,Desktops Computers,All-purpose,"<ul> <li>Model : HP Compaq 3330 Pro MT + 19"" L...",1259.0,local,0.12018,0.50220


### Text Cleaning & Vectorization

A custom function clean_text():

Removes HTML tags using BeautifulSoup

Eliminates special characters and digits

Converts text to lowercase

Strips extra whitespace

This function is applied to both the product title and description for the training and validation sets.

The cleaned title and description are concatenated to form a new feature called text.

Then:

A character-level n-gram vectorizer is defined using CountVectorizer with ngram_range=(2, 6) and max_features=5000, as outlined in the original paper.

The vectorizer is fitted on the training text, and both training and validation texts are transformed into numerical feature vectors.

Finally, the target variables y_clarity and y_conciseness are extracted for model training.

In [9]:
def clean_text(text):
    if pd.isna(text):
        return ""
    #Remove HTML tags
    text = BeautifulSoup(text, "html.parser").get_text()
    #Remove special characters and digits
    text = re.sub(r"[^a-zA-Z\s]", "", text)
    #Convert to lowercase
    text = text.lower()
    #Remove extra spaces
    text = re.sub(r"\s+", " ", text).strip()
    return text

#Applying transformations  to both train and validation sets
for df in [train, validation]:
    df['clean_title'] = df['title'].apply(clean_text)
    df['clean_desc'] = df['description'].apply(clean_text)
    df['text'] = df['clean_title'] + " " + df['clean_desc']

In [10]:
#Using character n-grams as described in the paper (ngram_range = (2,6))
char_vectorizer = CountVectorizer(analyzer='char', ngram_range=(2, 6), max_features=5000)

#Fitting on train text, transform train and validation
X_train = char_vectorizer.fit_transform(train['text'])
X_val = char_vectorizer.transform(validation['text'])

#Targets
y_clarity = train['clarity']
y_conciseness = train['conciseness']

### Bagged Ensemble Model Training with LightGBM

This section implements a robust ensemble strategy to improve prediction accuracy by reducing variance:

Model Choice:

Used LightGBM Regressor (LGBMRegressor) for its speed and performance on high-dimensional, sparse data (like character n-grams).

Ensemble Strategy – Bagging + K-Fold CV:

Bagging (n_bags=4): Run the full 10-fold cross-validation process four times with different random seeds to simulate different “bags” of models.

K-Fold Cross-Validation (n_folds=10): Each bag uses 10-fold CV to split the training data into training and validation subsets.

In total, 40 models are trained (4 bags × 10 folds).

Predictions from all models are averaged to produce final validation predictions.

Model Configuration:

n_estimators=300: Each model trains up to 300 boosting iterations.

learning_rate=0.05: Slower learning for better generalization.

num_leaves=31: Controls model complexity.

random_state: Varied per bag to encourage diversity in learned models.

Data Compatibility Fix:

Transformed X_train and X_val to np.float32 to resolve type compatibility errors with LightGBM (which expects float32 or float64 input, not int64).

Target-wise Training:

The process is run independently for both target variables:

clarity

conciseness

This setup helps create more stable and generalizable predictions by aggregating results from a diverse set of models trained on different subsets of data.

In [11]:
def run_lgbm_bagged(X, y, X_val, n_bags=4, n_folds=10):
    val_preds = np.zeros(X_val.shape[0])
    
    for bag in range(n_bags):
        kf = KFold(n_splits=n_folds, shuffle=True, random_state=42 + bag)
        for train_idx, valid_idx in tqdm(kf.split(X), desc=f"Bag {bag+1}/{n_bags}"):
            X_train_kf, X_valid_kf = X[train_idx], X[valid_idx]
            y_train_kf, y_valid_kf = y.iloc[train_idx], y.iloc[valid_idx]
            
            model = lgb.LGBMRegressor(
                n_estimators=300,
                learning_rate=0.05,
                max_depth=-1,
                num_leaves=31,
                random_state=42
            )
            model.fit(X_train_kf, y_train_kf)
            val_preds += model.predict(X_val) / (n_folds * n_bags)
    
    return val_preds


# Fix dtype for LightGBM
X_train = X_train.astype(np.float32)
X_val = X_val.astype(np.float32)

# Run for both targets
print("Training ensemble for CLARITY...")
val_preds_clarity = run_lgbm_bagged(X_train, y_clarity, X_val)

print("\nTraining ensemble for CONCISENESS...")
val_preds_conciseness = run_lgbm_bagged(X_train, y_conciseness, X_val)

Training ensemble for CLARITY...


Bag 1/4: 0it [00:00, ?it/s]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.687818 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36642
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943315


Bag 1/4: 1it [00:13, 13.10s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.696697 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36547
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943498


Bag 1/4: 2it [00:25, 12.90s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.567349 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36533
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943407


Bag 1/4: 3it [00:37, 12.16s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.596520 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36484
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943623


Bag 1/4: 4it [00:48, 11.75s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.581146 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36555
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.942949


Bag 1/4: 5it [00:59, 11.57s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.603670 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36579
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943041


Bag 1/4: 6it [01:11, 11.82s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.539506 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36568
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943531


Bag 1/4: 7it [01:23, 11.77s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.548980 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36588
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943225


Bag 1/4: 8it [01:35, 11.79s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.570489 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36336
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943255


Bag 1/4: 9it [01:47, 11.94s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.565744 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36654
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943776


Bag 1/4: 10it [01:59, 11.93s/it]
Bag 2/4: 0it [00:00, ?it/s]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.610368 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36480
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943376


Bag 2/4: 1it [00:12, 12.56s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.553069 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36593
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943774


Bag 2/4: 2it [00:24, 12.10s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.567243 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36604
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.942794


Bag 2/4: 3it [00:36, 11.94s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.551975 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36595
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943225


Bag 2/4: 4it [00:48, 11.96s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.593954 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36642
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943010


Bag 2/4: 5it [01:01, 12.35s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.567827 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36328
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.944021


Bag 2/4: 6it [01:13, 12.28s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.559027 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36369
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943653


Bag 2/4: 7it [01:25, 12.29s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.606005 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36655
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943837


Bag 2/4: 8it [01:38, 12.60s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.592184 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36602
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.942551


Bag 2/4: 9it [01:51, 12.72s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.686167 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36564
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943378


Bag 2/4: 10it [02:04, 12.43s/it]
Bag 3/4: 0it [00:00, ?it/s]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.558274 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36540
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943009


Bag 3/4: 1it [00:12, 12.29s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.554420 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36543
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943621


Bag 3/4: 2it [00:26, 13.20s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.757569 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36609
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.942855


Bag 3/4: 3it [00:41, 14.18s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.788163 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36589
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943960


Bag 3/4: 4it [00:57, 14.90s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.861597 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36596
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.942918


Bag 3/4: 5it [01:13, 15.22s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.959622 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36607
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.942980


Bag 3/4: 6it [01:30, 15.78s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.690769 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36507
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.944082


Bag 3/4: 7it [01:44, 15.17s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.625337 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36442
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943439


Bag 3/4: 8it [01:57, 14.56s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.611297 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36489
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943776


Bag 3/4: 9it [02:10, 14.04s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.585265 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36563
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.942980


Bag 3/4: 10it [02:23, 14.32s/it]
Bag 4/4: 0it [00:00, ?it/s]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.561350 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36583
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.942917


Bag 4/4: 1it [00:12, 12.97s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.793463 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36601
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943254


Bag 4/4: 2it [00:26, 13.49s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.576424 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36569
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943437


Bag 4/4: 3it [00:39, 13.13s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.683953 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36542
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943500


Bag 4/4: 4it [00:53, 13.55s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.833634 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36329
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.942704


Bag 4/4: 5it [01:07, 13.79s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.639161 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36535
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943592


Bag 4/4: 6it [01:21, 13.57s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.677781 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36483
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943898


Bag 4/4: 7it [01:37, 14.35s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.794079 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36547
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943439


Bag 4/4: 8it [01:52, 14.83s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.561767 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36611
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943225


Bag 4/4: 9it [02:05, 14.12s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.735766 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36587
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.943653


Bag 4/4: 10it [02:18, 13.86s/it]



Training ensemble for CONCISENESS...


Bag 1/4: 0it [00:00, ?it/s]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.610655 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36642
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685245


Bag 1/4: 1it [00:13, 13.75s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.704785 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36547
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.686103


Bag 1/4: 2it [00:28, 14.36s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.621618 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36533
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685888


Bag 1/4: 3it [00:42, 14.17s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.703180 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36484
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685132


Bag 1/4: 4it [00:57, 14.33s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.722448 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36555
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685041


Bag 1/4: 5it [01:12, 14.64s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.601804 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36579
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.684214


Bag 1/4: 6it [01:26, 14.47s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.579901 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36568
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.684459


Bag 1/4: 7it [01:40, 14.36s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.685947 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36588
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685592


Bag 1/4: 8it [01:55, 14.66s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.586680 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36336
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685684


Bag 1/4: 9it [02:09, 14.43s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.774596 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36654
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685990


Bag 1/4: 10it [02:24, 14.44s/it]
Bag 2/4: 0it [00:00, ?it/s]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.917905 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36480
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.684694


Bag 2/4: 1it [00:16, 16.39s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.985660 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36593
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685950


Bag 2/4: 2it [00:32, 16.26s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.599121 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36604
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685276


Bag 2/4: 3it [00:48, 15.89s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.764127 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36595
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685439


Bag 2/4: 4it [01:02, 15.48s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.643323 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36642
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.686051


Bag 2/4: 5it [01:16, 14.91s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.565563 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36328
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.686449


Bag 2/4: 6it [01:29, 14.05s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.631795 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36369
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.684612


Bag 2/4: 7it [01:42, 13.93s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.604530 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36655
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685622


Bag 2/4: 8it [01:56, 13.74s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.569170 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36602
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.686480


Bag 2/4: 9it [02:08, 13.39s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.653678 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36564
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.682774


Bag 2/4: 10it [02:22, 14.22s/it]
Bag 3/4: 0it [00:00, ?it/s]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.604402 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36540
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.683990


Bag 3/4: 1it [00:12, 12.93s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.693707 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36543
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.686225


Bag 3/4: 2it [00:26, 13.49s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.571682 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36609
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685705


Bag 3/4: 3it [00:40, 13.38s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.622708 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36589
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685102


Bag 3/4: 4it [00:53, 13.36s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.626149 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36596
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685010


Bag 3/4: 5it [01:06, 13.32s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.550728 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36607
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.684887


Bag 3/4: 6it [01:19, 13.29s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.556772 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36507
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685224


Bag 3/4: 7it [01:32, 13.12s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.527600 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36442
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685347


Bag 3/4: 8it [01:47, 13.83s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.741144 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36489
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.686694


Bag 3/4: 9it [02:01, 13.69s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.529240 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36563
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685163


Bag 3/4: 10it [02:15, 13.55s/it]
Bag 4/4: 0it [00:00, ?it/s]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.899249 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36583
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685582


Bag 4/4: 1it [00:16, 16.83s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.765854 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36601
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.684970


Bag 4/4: 2it [00:30, 15.19s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.635619 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36569
[LightGBM] [Info] Number of data points in the train set: 32654, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685429


Bag 4/4: 3it [00:44, 14.48s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.558076 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36542
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.684520


Bag 4/4: 4it [00:57, 13.97s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.549807 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36329
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.684244


Bag 4/4: 5it [01:10, 13.55s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.689048 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36535
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.686266


Bag 4/4: 6it [01:24, 13.58s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.544802 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36483
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685408


Bag 4/4: 7it [01:36, 13.28s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.550020 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36547
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685132


Bag 4/4: 8it [01:49, 13.02s/it]

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.582132 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 36611
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.686357


Bag 4/4: 9it [02:02, 12.95s/it]

[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.606780 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 36587
[LightGBM] [Info] Number of data points in the train set: 32655, number of used features: 5000
[LightGBM] [Info] Start training from score 0.685439


Bag 4/4: 10it [02:15, 13.57s/it]


In [12]:
#Extracting actual labels from validation set
y_val = validation[['clarity', 'conciseness']]

#Calculate RMSE
rmse_clarity = mean_squared_error(y_val['clarity'], val_preds_clarity, squared=False)
rmse_conciseness = mean_squared_error(y_val['conciseness'], val_preds_conciseness, squared=False)

print(f"Validation RMSE - Clarity: {rmse_clarity:.5f}")
print(f"Validation RMSE - Conciseness: {rmse_conciseness:.5f}")

Validation RMSE - Clarity: 0.52931
Validation RMSE - Conciseness: 0.35273


### Results Comparison: Our Implementation vs. Research Paper

This section compares our model’s validation performance with the results reported in the CIKM AnalytiCup 2017 paper titled "Lazada Product Title
Quality Challenge Bagging Model for Product Title Quality with Noise".

Target	     Algorithm	   Paper RMSE	  Our RMSE
Conciseness	 XGBoost	      0.31553	   0.35273
Clarity	     XGBoost	      0.20745	   0.52931

Observations:
Our RMSE values are higher, particularly for the Clarity task.

The paper’s results were achieved using additional techniques such as:

Ensemble of different algorithms (e.g., XGBoost, Ridge, SVR)

More aggressive hyperparameter tuning

Possibly more preprocessing steps or feature engineering

Our pipeline is simpler (using only LightGBM + character n-grams), but still reasonably competitive for conciseness.

# Transformer Approach

In [13]:
!pip install transformers torch

Defaulting to user installation because normal site-packages is not writeable


In [14]:
import torch
from transformers import DistilBertTokenizer, DistilBertModel
from sklearn.linear_model import Ridge

# 1. Define a separate cleaner for Transformers
# We keep numbers and punctuation because BERT understands context like "50ml" or "100%"
def transformer_clean_text(text):
    if pd.isna(text):
        return ""
    # Remove HTML tags only
    text = BeautifulSoup(text, "html.parser").get_text()
    # Remove extra white spaces
    text = re.sub(r"\s+", " ", text).strip()
    # We DO NOT remove special chars or numbers here
    return text

# 2. Apply new cleaning to a new column to avoid messing up your existing 'text' column
print("Preparing text for Transformer...")
for df in [train, validation]:
    df['trans_text'] = (df['title'].apply(transformer_clean_text) + 
                        ". " + 
                        df['description'].apply(transformer_clean_text))

Preparing text for Transformer...


In [15]:
# Initialize Tokenizer and Model
# 'distilbert-base-uncased' is smaller and faster than full BERT, perfect for CPU
tokenizer = DistilBertTokenizer.from_pretrained('distilbert-base-uncased')
model = DistilBertModel.from_pretrained('distilbert-base-uncased')

# Function to generate embeddings in batches
def get_bert_embeddings(text_list, batch_size=32):
    model.eval() # Set model to evaluation mode
    embeddings = []
    
    # Process in batches to show progress and save RAM
    for i in tqdm(range(0, len(text_list), batch_size), desc="Extracting Embeddings"):
        batch = text_list[i : i + batch_size]
        
        # Tokenize: Pad to max length in batch, truncate to 512 (BERT limit)
        inputs = tokenizer(list(batch), return_tensors="pt", padding=True, truncation=True, max_length=512)
        
        with torch.no_grad(): # Disable gradient calculation for speed/memory
            outputs = model(**inputs)
        
        # We take the embedding of the [CLS] token (first token) which represents the whole sentence
        # Shape: [batch_size, 768]
        batch_embeddings = outputs.last_hidden_state[:, 0, :].numpy()
        embeddings.append(batch_embeddings)
        
    return np.vstack(embeddings)

# Generate Embeddings (This might take 10-15 mins on CPU depending on dataset size)
print("Extracting embeddings for Training set...")
X_train_bert = get_bert_embeddings(train['trans_text'].tolist())

print("Extracting embeddings for Validation set...")
X_val_bert = get_bert_embeddings(validation['trans_text'].tolist())

print(f"Embedding Shape: {X_train_bert.shape}")

Extracting embeddings for Training set...


Extracting Embeddings: 100%|██████████| 1134/1134 [1:13:27<00:00,  3.89s/it]


Extracting embeddings for Validation set...


Extracting Embeddings: 100%|██████████| 370/370 [25:35<00:00,  4.15s/it]

Embedding Shape: (36283, 768)


In [16]:
# 1. Train for CLARITY
print("Training Transformer-based Ridge for Clarity...")
ridge_clarity = Ridge(alpha=1.0)
ridge_clarity.fit(X_train_bert, y_clarity)
bert_preds_clarity = ridge_clarity.predict(X_val_bert)

# 2. Train for CONCISENESS
print("Training Transformer-based Ridge for Conciseness...")
ridge_conciseness = Ridge(alpha=1.0)
ridge_conciseness.fit(X_train_bert, y_conciseness)
bert_preds_conciseness = ridge_conciseness.predict(X_val_bert)

# 3. Evaluation
rmse_bert_clarity = mean_squared_error(y_val['clarity'], bert_preds_clarity, squared=False)
rmse_bert_conciseness = mean_squared_error(y_val['conciseness'], bert_preds_conciseness, squared=False)

print("-" * 30)
print("RESULTS COMPARISON (RMSE)")
print("-" * 30)
print(f"Original LightGBM (Char N-gram) - Clarity:    {rmse_clarity:.5f}")
print(f"Transformer (DistilBERT + Ridge) - Clarity:   {rmse_bert_clarity:.5f}")
print("-" * 30)
print(f"Original LightGBM (Char N-gram) - Conciseness: {rmse_conciseness:.5f}")
print(f"Transformer (DistilBERT + Ridge) - Conciseness: {rmse_bert_conciseness:.5f}")

Training Transformer-based Ridge for Clarity...
Training Transformer-based Ridge for Conciseness...
------------------------------
RESULTS COMPARISON (RMSE)
------------------------------
Original LightGBM (Char N-gram) - Clarity:    0.52931
Transformer (DistilBERT + Ridge) - Clarity:   0.53035
------------------------------
Original LightGBM (Char N-gram) - Conciseness: 0.35273
Transformer (DistilBERT + Ridge) - Conciseness: 0.35992


Blending the two predictions. Since LightGBM was better, we will give it more weight (e.g., 70% LightGBM, 30% BERT).

In [17]:
print("--- ENSEMBLING PREDICTIONS ---")

# We assign weights based on performance. 
# Since LightGBM was better, we give it higher weight (0.7).
w_lgb = 0.7
w_bert = 0.3

# Blend Clarity
ensemble_preds_clarity = (w_lgb * val_preds_clarity) + (w_bert * bert_preds_clarity)
rmse_ens_clarity = mean_squared_error(y_val['clarity'], ensemble_preds_clarity, squared=False)

# Blend Conciseness
ensemble_preds_conciseness = (w_lgb * val_preds_conciseness) + (w_bert * bert_preds_conciseness)
rmse_ens_conciseness = mean_squared_error(y_val['conciseness'], ensemble_preds_conciseness, squared=False)

print(f"Ensemble RMSE - Clarity:     {rmse_ens_clarity:.5f} (Improvement over best: {rmse_clarity - rmse_ens_clarity:.5f})")
print(f"Ensemble RMSE - Conciseness: {rmse_ens_conciseness:.5f} (Improvement over best: {rmse_conciseness - rmse_ens_conciseness:.5f})")

--- ENSEMBLING PREDICTIONS ---
Ensemble RMSE - Clarity:     0.52924 (Improvement over best: 0.00006)
Ensemble RMSE - Conciseness: 0.35274 (Improvement over best: -0.00001)


# Conclusion & Model Comparison

In this project, we approached the CIKM AnalytiCup challenge (predicting Clarity and Conciseness scores) using three distinct strategies. Below is a comparison of their performance (RMSE). [LOWER IS BETTER]

| Technique | Clarity RMSE | Conciseness RMSE | Key Observation |
| :--- | :--- | :--- | :--- |
| **1. Baseline: Char N-Grams + LightGBM** | **0.52931** | **0.35273** | **Best Single Model.** Character N-grams (2-6) were highly effective at capturing structural patterns (spelling, capitalization, punctuation) which correlate strongly with clarity scores. |
| **2. Transformer: DistilBERT + Ridge** | 0.53035 | 0.35992 | **Slightly Worse.** While DistilBERT captures deep semantic meaning, the "frozen" embeddings combined with a linear model (Ridge) struggled to outperform the specific structural signals captured by N-grams. |
| **3. Ensemble: Weighted Average** | **0.52924** | 0.35274 | **Marginal Improvement.** Blending the models (70% Baseline, 30% Transformer) provided a tiny gain in Clarity (0.00007), suggesting that while the Transformer adds some diversity, the Baseline is already capturing most of the signal. |

### Analysis
* **Why N-grams won:** "Clarity" and "Conciseness" in e-commerce are often about *formatting* (e.g., bullet points, length, number of specs) rather than deep semantic meaning. Character N-grams capture these "visual" text features better than semantic embeddings.
* **Transformer Limitation:** Since we used **Feature Extraction** (frozen weights) due to compute constraints, we could not fine-tune the BERT model to specifically look for "clarity." A fine-tuned RoBERTa model would likely outperform the baseline if a GPU were available.